### PR TITLE
Orv2 1501: added endpoint for search company by company name or client id

### PR DESCRIPTION
**Pull request by bcgov-brwang**
# Description

Please provide a summary of the change and the issue fixed. Please include relevant context. List dependency changes.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist

- [ ] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have already been accepted and merged

## Further comments

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://onroutebc-1069-frontend.apps.silver.devops.gov.bc.ca)
- [Vehicles](https://onroutebc-1069-vehicles.apps.silver.devops.gov.bc.ca/api)
- [Dops](https://onroutebc-1069-dops.apps.silver.devops.gov.bc.ca/api)
- [TPS-Migration](https://onroutebc-1069-tps-migration.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/onroutebc/actions/workflows/analysis.yml)

After merge, new images are promoted to:
- [Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge.yml)

### CODE REVIEW

File: vehicles/src/modules/company-user-management/company/company.controller.ts [hunk 1]
```diff
@@ -81,6 +84,39 @@ export class CompanyController {
...
+  @ApiQuery({ name: 'legalName', required: false })
+  @ApiQuery({ name: 'clientNumber', required: false })
+  @Roles(Role.READ_ORG)
+  @Get('paginated')
+  async getCompanyMetadataPaginated(
+    @Query() pageOptionsDto: PageOptionsDto,
+    @Query('legalName') legalName: string,
+    @Query('clientNumber') clientNumber: string,
+  ): Promise<PaginationDto<ReadCompanyMetadataDto>> {
+    console.log("Legal Name passed in:", legalName);
```
Comment (erikataot) [inline vehicles/src/modules/company-user-management/company/company.controller.ts]:
```suggestion

```

File: vehicles/src/modules/company-user-management/company/company.controller.ts [hunk 2]
```diff
@@ -81,6 +84,39 @@ export class CompanyController {
...
+    const companies: PaginationDto<ReadCompanyMetadataDto> =
+      await this.companyService.findCompanyMetadataPaginated(pageOptionsDto, legalName, clientNumber);
+
+    if (!companies?.items.length) {
```
Comment (erikataot) [inline vehicles/src/modules/company-user-management/company/company.controller.ts]:
```suggestion
    if (!companies?.items?.length) {
```

File: vehicles/src/modules/company-user-management/company/company.service.ts [hunk 3]
```diff
@@ -282,6 +286,46 @@ export class CompanyService {
...
+   * @param clientNumber The company's client number
+   * @returns The company list as a promise of type {@link ReadCompanyMetadataDto}
+   */
+  @LogAsyncMethodExecution()
+  async findCompanyMetadataPaginated(
+    @Query() pageOptionsDto: PageOptionsDto,
+    @Query() legalName: string,
+    @Query() clientNumber: string,
+  ):Promise<PaginationDto<ReadCompanyMetadataDto>> {
+    console.log("Legal Name received:", legalName);
```
Comment (erikataot) [inline vehicles/src/modules/company-user-management/company/company.service.ts]:
```suggestion

```

File: vehicles/src/modules/company-user-management/company/company.service.ts [hunk 4]
```diff
@@ -282,6 +286,46 @@ export class CompanyService {
...
+      })
+      .getMany();
+
+    const companyMetadata = await this.classMapper.mapArrayAsync(
+      companies,
+      Company,
+      ReadCompanyMetadataDto,
+    );
+
+    const totalItems = companyMetadata.length;
```
Comment (erikataot) [inline vehicles/src/modules/company-user-management/company/company.service.ts]:
```suggestion
    const totalItems = companyMetadata?.length;
```

File: vehicles/test/unit/company/company.controller.spec.ts [hunk 5]
```diff
@@ -160,4 +161,31 @@ describe('CompanyController', () => {
...
+    it('should return the company metadata when company legal name or client id are provided', async () => {
+      const request = createMock<Request>();
+      request.user = redCompanyAdminUserJWTMock;
+      const pageOptionsDto = {
+        page: 1,
+        take: 10,
+      }
+      const retCompanyMetadata = await controller.getCompanyMetadataPaginated(pageOptionsDto, 'He', 'R');
+      expect(typeof retCompanyMetadata).toBe('object');
+      console.log("Retrieved records: " + retCompanyMetadata.items.length.toString());
```
Comment (erikataot) [inline vehicles/test/unit/company/company.controller.spec.ts]:
```suggestion

```

File: vehicles/src/modules/company-user-management/company/company.service.ts (hunk 4 above)
Comment (bcgov-brwang) [reply vehicles/src/modules/company-user-management/company/company.service.ts]: Updated as suggested.

**Review by praju-aot — APPROVED** [review]
Approved

**Pull request merged**